### PR TITLE
fix: exitCodes, reporters and bail, better error logging

### PR DESCRIPTION
**Pull request by shellscape**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case

It's been reported that #35 didn't do enough to address exit codes in the event of a build failure or error. During discovery, it was also found that the `--bail` flag will cause errors in the reporters as well, as they were not anticipating `stats` being `undefined` in the event that `--bail` was used. 

Additionally, the logic for how an error is displayed when one occurs at the program level (outside of reporters) has been improved greatly and gives users more options to display additional info.

### Breaking Changes

None

### Additional Info

/cc @ankurk91 @MrMamen

### CODE REVIEW

**Comment by ankurk91** [comment]
Thanks for covering the `--bail` issue. I was about to report that.

This  PR also might fix -https://github.com/webpack-contrib/webpack-command/issues/31

File: lib/compiler.js [hunk 1]
```diff
@@ -29,6 +35,10 @@ function makeCallback(options) {
     if (error) {
       reject(error);
       return;
+    } else if (stats && stats.hasErrors() && !watch) {
+      process.exitCode = getErrorCount(stats);
```
Comment (ooflorent) [inline lib/compiler.js]: Please use `1` instead. Using the error count as the exit code when prevent _us_ from using meaningful non-zero exit codes.

Comment (MrMamen) [reply lib/compiler.js]: https://people.cs.pitt.edu/~alanjawi/cs449/code/shell/UnixSignals.htm

Comment (evilebottnawi) [reply lib/compiler.js]: Really looks strange

Comment (shellscape) [reply lib/compiler.js]: resolved

**Pull request merged**